### ISSUE TITLE
Aborting a script with Ctrl-C doesn't actually kill jobs using qdel

### ISSUE THREAD

**Issue by golharam** [pcingola/BigDataScript#74]
Hi - Recently I noticed with a new version (since .d) anytime I press Ctrl-C to abort a script, I see BDS issue qdel commands, however it looks like the underlying jobs on the compute nodes are still running.  Its almost as if BDS die but doesn't kill the tasks it has spawned.

**Comment by golharam** [comment]
My sys admins insist nothing changed on our servers.  The only other change has been an updated BDS.  So I'm not sure where the cause is just yet but wanted to see if anything in BDS changed that could be causing this.

**Comment by pcingola** [comment]
If you issue exactly the same 'qdel' command by hand, do the jobs 
get killed? (the commands are shown in the "pid" file before you Ctrl-C).

**Comment by golharam** [comment]
Yes, I have a single job submitted by BDS...its a long running job.  I
invoke BDS with my script, I see the job in qstat and I see the job running
on a compute node.  I manually issue a qdel with the job id.  I see the job
removed in qstat but when I run top on the compute node, I see bds is will
running that particular task.

**Comment by pcingola** [comment]
I don't think I understand what you mean. Is the problem that qdels are 
not issued by bds or that when you manually kill the job bds keeps running?

For the first part: bds logs the command that has to use to kill the remaining tasks 
in a file called "bds.*.pid". When you press Ctrl-C, the Java part of the program 
dies immediately and the low level part (written in Go) captures the signal and 
executes de commands in that "pid" file. The question was: If you execute those 
'qdel' commands, do they kill the task? (if there is a problem, I'm guessing it might 
be there). I'll try to debug entries to show the commands, so it's easier to see.

For the second part: If you manually kill the job, bds will see as "disappeared from 
the cluster" but it won't be immediately. Bds monitors the cluster every minute or so 
and each time the job is missing, a counter is incremented. When the counter 
reaches a certain value (default 3) then the task is marked as "missing" and then 
as "failed".

**Comment by pcingola** [comment]
OK, I double checked and it turns out that bds is actually logging 'qdel' commands 
correctly.

Here is an example:

    $ cat z.bds
    task echo Running in cluster; sleep 600; echo Finished running in cluster

    $ bds -s cluster -d z.bds 
    ...  
    00:00:00.794	TaskLogger: Appending to PidFile 'bds.pid.28031', lines:
                            |27671559.gm-1r16-n04	+	qdel

So if you go to another terminal and check, you'll see the task in 
the cluster's queue:

    $ qstat | grep 27671559.gm-1r16-n04
    27671559.gm-1r16-n04       STDIN            pcingola               0 Q serial-short   

Now, if I go back to the first terminal (where bds is running) and 
press Ctrl-C

    $ qstat | grep 27671559.gm-1r16-n04
    27671559.gm-1r16-n04       STDIN            pcingola               0 C serial-short   

as you can see, the job reported by qstat changed from Q (queued) 
to C (cancelled). Everything seems to be working OK.

**Comment by pcingola** [comment]
This is an example for the "disappearing task" part of your question.
Running the same script as before:

    $ cat z.bds
    task echo Running in cluster; sleep 600; echo Finished running in cluster

    $ bds -s cluster -d z.bds 
    ...
    00:00:00.790    TaskLogger: Appending to PidFile 'bds.pid.25480', lines:
                     |27672780.gm-1r16-n04   +       qdel

Now I go to another terminal and remove the job form the cluster 
using 'qdel'

    $ qdel 27672780.gm-1r16-n04
    
    $ qstat | grep 27672780.gm-1r16-n04
    27672780.gm-1r16-n04       STDIN            pcingola               0 C serial-short

Now the job has been "Cancelled" as expected. Nevertheless in 
this cluster, cancelled jobs remain in the queue roughly for 10 minutes.
So after 11 minutes, the job disappears from the cluster's queue. An bds
starts incrementing the "missing job" counter:

    00:11:16.082    WARNING: Task PID '27672780.gm-1r16-n04' not found for task 'z.bds.20150224_165011_928/task.z.line_3.id_1'. Incrementing 'missing counter': 1 (max. allowed 3)

This happens until the counter is more than 3:

    00:12:17.457    WARNING: Task PID '27672780.gm-1r16-n04' not found for task 'z.bds.20150224_165011_928/task.z.line_3.id_1'. Incrementing 'missing counter': 2 (max. allowed 3)
    ...
    00:13:18.969    WARNING: Task PID '27672780.gm-1r16-n04' not found for task 'z.bds.20150224_165011_928/task.z.line_3.id_1'. Incrementing 'missing counter': 3 (max. allowed 3)
    ...
    00:14:19.831    WARNING: Task PID '27672780.gm-1r16-n04' not found for task 'z.bds.20150224_165011_928/task.z.line_3.id_1'. Incrementing 'missing counter': 4 (max. allowed 3)

Now the task is marked as failed missing and then as failed:

        Task PID '27672780.gm-1r16-n04' not found. Marking it as finished.

    Task failed:
            Program & line     : 'z.bds', line 3
            Task Name          : ''
            Task ID            : 'z.bds.20150224_165011_928/task.z.line_3.id_1'
            Task PID           : '27672780.gm-1r16-n04'


  So, the task is detected to be missing, but it takes roughly 14 minutes (other 
clusters may have different time settings on how long they keep the tasks).
Disappearing jobs are quite unusual, unless your cluster is having severe 
problems.
  I hope this clarifies how the mechanism works.

**Comment by kmhernan** [comment]
If I submit my bds script in a job script (e.g., sge) so I don't have to run bds on the head node, or use something like `screen` and I decide I want to stop the BDS script, using `qdel` to delete the job running my master BDS script will stop that job, but BDS will not kill any of the other jobs. Is there any way to get around this, or any suggestions??

**Comment by pcingola** [comment]
If you want to stop the pipeline, the correct way would be to send a kill signal to 
the bds process (like you do with any "normal" program). This lets bds do the cleanup.
You should use a kill -HUP (because a "kill -9" will not allow for cleanup).
BDS saves a file called 'bds.NNNN.pid' to make it easier to track the pid number.

Now, if you don't have access to the node where bds is running...well, honestly I never 
really thought of that case because I always run bds on the head node/s and I never 
had this problem. If this is a common situation for you, we can try to think some 
alternatives...

**Comment by kmhernan** [comment]
So, I'm currently using the sge script method because some admins get angry when things are run on the head node and telling someone to `qsub file.sge` rather than teach them about `screen` is somewhat easier. So if possible, I'd like to think of some alternatives... Let me know

**Comment by pcingola** [comment]
Hi Kyle,
  I totally understand that admins don't want to run "data processing" programs on master 
nodes and it makes total sense. Nevertheless, running "pipeline control" software
in the master-nodes is OK in almost every cluster I've seen (that is actually the 
reason why master-nodes exist). 
   Remember running bds is conceptually similar to running a shell script that 
sends qsub commands. I assume your administrators would not oppose to running 
a shell script with the same functionality, would they?
   I'll think about what you say, but this looks more like an "overzealous admin" 
problem, rather than a bds problem....

**Comment by kmhernan** [comment]
haha I can agree.. and I believe I could get away with it myself... So, how about, if I were to right just a bash script that ran the bds pipeline? If I were to `ctrl-C` while running the bash script, should I expect the child jobs to be qdel'd? I think I should, just want to check.

**Comment by pcingola** [comment]
It's better to just sun bds directly, but yes, I would expect jobs to be cleaned up when Ctrl-C on a bash script.

**Comment by kmhernan** [comment]
So, some testing yesterday seemed to show that ctrl-c on a bash script sometimes worked... but sometimes java jobs were still left hanging... I didn't try kill... I can try today

**Comment by golharam** [comment]
Same here.  I see my java jobs still running.

**Comment by pcingola** [comment]
So, you run bds, press Ctrl-C and the jobs in the nodes do not die?
Can you send a simple script that I can try? What cluster type are you using?

Personally, I use Ctrl-C all the time and bds cleans up everything 
for me (I never had issues so far).

**Comment by golharam** [comment]
Yes, I used 'bds -s sge myscript.bds"

I'm using SGE.  Attached is a copy of a script that does this with bwa

This is a subset of my script.  Subsequent tasks depend on this task.
After the script submits these tasks, it waits for them to complete.  While
waiting, I press Ctrl-C to abort.  I see the jobs get deleting via qstat,
but when I log into the compute nodes and run top, I see bwa is still
running.

**Comment by pcingola** [comment]
OK, so if your qstat does not show the job any more, then BDS did 
clean up the tasks by invoking qdel. 
  It looks like the bug is in your cluster management system. After BDS 
sent a 'qdel' and the job was properly de-allocated (i.e. you don't see 
it when you run 'qstat'), but the process is not cleaned up from the 
node.
   BDS delegates clean up to the cluster (it does not have any way 
of doing it itself after qdel is invoked). So it doesn't seem to be a 
BDS issue (please do correct me if you think I'm wrong).

**Comment by kmhernan** [comment]
I think you're right @pcingola ... I haven't had this issue since I last posted and it would make sense that if you no longer see the task in `qstat` but you can still see the tool running with `top`, then it is SGE's issue. The only other thing would be if somehow some tasks were running local.

**Comment by pcingola** [comment]
If you have the issue with tasks running locally, please let me know (so far it never happened to me).
When tasks are local, bds issues a 'kill' instead of a qdel.

**Comment by golharam** [comment]
I need to prove this to our IT folks.  How does BDS work on the compute
node?  It looks like BDS is invoked on the compute node and runs the
underlying task.  So when SGE tries to kill the job, its actually trying to
kill bds on the compute node.  Is this correct?

I need to show a shell script that calls java, then qdel the shell script.
If I still see java running on the compute node, then they will buy this.
If java is killed, then they will point to BDS not killing the java process.

**Comment by pcingola** [comment]
BDS is called in the compute node using 'bds -exec' which executes 
the underlying command (as a child process), enforces time limits, redirects 
output, collects exit codes, etc. Take a look at Figure 2 in this paper (it 
might help to clarify how this works): 
http://bioinformatics.oxfordjournals.org/content/31/1/10.full.pdf?keytype=ref&ijkey=ecsDSVYrtkLMMpN

All this has to be done because many cluster systems either do not have 
such features or they don't work (usually because they are misconfigured) or 
have weird behaviours than cannot be reliably predicted across different 
cluster or systems (e.g. processes executed in remote 'ssh' host do not 
have any of those features).

In a nutshell, if the cluster is killing the parent process and not cleaning 
the child process correctly, it may kill only 'bds -exec' and not the 
child 'bwa' process. 

If your cluster cannot be configured to correctly clean the child processes, let 
me know and we'll think a way to make this work...

**Comment by kmhernan** [comment]
@pcingola I am seeing this exact issue using Amazon EC2 and MIT's star cluster... do you know anything about setting up the `qdel` to kill child processes? I'm doing lots of googling.

**Comment by pcingola** [comment]
This is weird. OK, I can test it on that platform...
I probably won't be able to spend much time on this  until next week, though (apologies)

**Comment by kmhernan** [comment]
@pcingola so after lots of hair-pulling, I see that `orte` is set up for openMPI and asserts that there is resource sharing across nodes and that there are control slaves. So, I had to create my own parallel environment that is more suited for openMP. In addition, I have added the option `ENABLE_ADDGRP_KILL=true` to the `execd_params` parameter of `qconf -sconf`. I'm not sure if this will solve the problems, but I haven't really seen the zombie tasks. I'm not an SGE export (nor do I want to be), so if someone else could chime in...

**Comment by pcingola** [comment]
A little bit of background: "bds -exec" is the GO program that wraps task execution in order 
to collect exit status, STDOUT/STDERR and enforce timeouts. Some systems/clusters cannot 
do these, that's why "bds -exec" is needed. When a task is executed, "bds- exec" needs to 
be able to kill all nested sub-processes (e.g. a shell that calls a command that calls a shell...). This 
can only be done by creating a new process group and sending the kill signal to the process 
group (otherwise the kill signal would kill the "bds -exec" process itself, which is not good).

After quite some googling myself, I think that indeed the 'ENABLE_ADDGRP_KILL=TRUE' is 
the best option for processes that create new process groups.

Meanwhile, I'll keep thinking on how to improve this situation. E.g. by "trusting" some 
cluster systems and not creating a new process group (this may create new side 
effects...

**Comment by golharam** [comment]
Any progress on this?  I can verify that if I launch of job via bds that
runs java, then press Ctrl-C to abort the job, the actually java process on
the compute node still runs.

If I submit the same java job as a shell script directly to SGE, then issue
a qdel, the java program is terminated.

**Comment by kmhernan** [comment]
@golharam can you verify if the  'ENABLE_ADDGRP_KILL=TRUE'  option is set on your SGE system?

**Comment by golharam** [comment]
It wasn't set.  I just added it, but this still occurs.

**Comment by golharam** [comment]
Actually, it may not have been set right.  Testing now

**Comment by golharam** [comment]
This appears to work now, so I'm closing this issue.
